fdbserver/storageserver.actor.cpp

```diff
@@ -3523,9 +3530,22 @@ ACTOR Future<Void> update(StorageServer* data, bool* pReceivedUpdate) {
 			if (data->otherError.getFuture().isReady())
 				data->otherError.getFuture().get();
 
+			data->fetchedVersions = ver - data->version.get();
+			data->duration = now() - data->lastUpdate;
```

Comment:
I haven't thought too carefully about this, but you could potentially turn `fetchedVersions` into a normal counter rather than a special one. The way I'm imagining this could work would be that you set the counter value to the current version each time you fetch it, and then the metric would be reported with the normal fields for a counter (rate per second, roughness, counter value). If you did it this way, you may not need the `duration` field because you are already getting a rate computed for you.

I suppose if you think about it, the counter value is just going to give you the version in that model. You could then basically just take the existing `Version` field in `StorageMetrics` and make it into a counter instead of a single number and get all of this data. That might be more controversial, so I'd be ok with the idea of making this a separate field.

Reply:
@xumengpanda preferred the current way of reporting over the "counter" based implementation. From the radar:

> 
> Which of the following better captures what is being asked in this issue?
> 
> - The rate at which versions are being fetched (not necessarily in the last “x” seconds)? This can be implemented by adding a “Counter” field that captures the number of versions that are being fetched and updating the counter in update() (in StorageServer.actor.cpp) and Status.actor.cpp reports the value returned by Counter::getRate().
> ....

Reply:
It looks to me like the quoted text is saying roughly the same thing as I did, which is to use the counter construct. I might be misunderstanding, though.

Reply:
Yes, it is kind of saying the same thing. I was trying to say @xumengpanda did not like this implementation (please check the comments in the radar).

Reply:
I don't think I have access to the radar

Reply:
Oh, I did not know that. Sorry about that. I will copy-paste here:

> 
> Which of the following better captures what is being asked in this issue?
> 
> - The rate at which versions are being fetched (not necessarily in the last “x” seconds)? This can be implemented by adding a “Counter” field that captures the number of versions that are being fetched and updating the counter in update() (in StorageServer.actor.cpp) and Status.actor.cpp reports the value returned by Counter::getRate().
> -  The number of versions fetched in the last “x” seconds (we will report both “x” and the number of versions)? This can be implemented by having a Counter that captures the number of versions that are being fetched and resetting this counter in update() (in StorageServer.actor.cpp) where “versionLag” is currently being updated. Status.actor.cpp reports the counter value and “x” seconds.
> -  The number of versions fetched in the last 5 seconds? This could be hard to implement as it requires that we need to maintain a moving window of 5 seconds and the versions fetched in that window. 
> - Some other way that is not captured by the above methods?
> 
> Meng, could you please clarify what is being asked here? Thanks!
> 
> It’s the second one

Reply:
Ok, I see. Having thought about this more, I think I like the idea of using the counter (option 1) the best, so we should maybe have @xumengpanda weigh in a bit more with his rationale so we can make an informed decision.

My reasons for preferring the counter based approach:

1. In the original formulation, I believe this was logging an event for every log update, which could potentially be a bit spammy. It was also in a non-standard location (i.e. not in `StorageMetrics`), which is marginally less discoverable. In the current formulation, it will essentially log roughly one sample of a log update every 5 seconds. Compared to logging the rate from a counter that accounts for all updates over the 5 second window, I think this is losing information.
2. In status, I think the more standard way of reporting this is as a rate rather than an absolute delta and a duration. That means we should have a good reason for preferring the latter if we do that. In either case, the value you get in status is going to be representative of a small window, but the 5s counter-based rate will capture more data than the point-in-time sample for the last update.

Reply:
my thought was: what if update() is starved -- not called for a long time. Can we catch that? 
with a single counter, we won't see it. We saw the rate is lower, but that could be due to tLog or SS is starved or the network is bad. 

I don't have a strong preference whether or not using counter. As long as we can answer the question with one query: 
how fast SS is catching up with tLog? and if it's slow, is it due to SS not update() frequently enough?

Reply:
I think in that case, having the counter is still preferable to reporting the last update for the following reasons:

1. You have 5 second resolution into the data, so you can easily see long stretches of no activity by looking at those periods where the rate was 0. In the current proposal it's always going to report the last update, so it may not be as obvious that nothing is happening (is the last update recent or from a long time ago?). You could make it so that it resets after being logged and reports a 0 for the period until the next event, but that seems pretty similar in behavior to the 5 second counter.
2. If you want to see how frequently update is being called, neither of these approaches is going to be great for that. Having a duration on the last event gives you the ability to guess the rate, and with enough samples you could probably estimate it reasonably well. Better yet, though, would be to track the update frequency itself as a counter and report the rate measured by it.
3. Having the rate from the counter doesn't lose information, so it should be better at answering the question "how fast are we catching up?".

So possibly we could capture both requirements by turning the existing `Version` field into a counter and adding a counter to track the number of times we fetch from the logs? Or you could not re-use the version field and add a separate one with the understanding that it's counter value should match the version.

Reply:
> If you want to see how frequently update is being called, neither of these approaches is going to be great for that.

StorageServer already has a field called "lastUpdate" which gets updated within "update()", maybe exposing that field in status (together with the information exposed in this PR) would help answer Meng's question?

Regarding the implementation choice (counter vs <value,duration>): If it is the rate that we want then we should go with using a counter but if it is the metrics about the latest batch (how many versions and how long it took to get them) then we should go with <value, duration>. Also, note that the source tLog Id that we are reporting in this PR is not going to be very relevant with the counter approach as it Is capturing the source tLog Id for the latest batch.

Reply:
I'm not sure if this is something that is planned to change, but assuming the SS is fairly sticky to one transaction log, the changes to your source log wouldn't have much impact on your metrics. You would just have a little uncertainty for one 5s window, but almost all windows would be attributable to one log. This is probably not substantially better with the value/duration pairs unless the log changes very frequently such that you could piece together a better story via the samples (note this requires joining with the source log change event).

> If it is the rate that we want then we should go with using a counter but if it is the metrics about the latest batch (how many versions and how long it took to get them) then we should go with <value, duration>

My gut says that having a rate is going to be more useful overall, and in particular is going to be better than a sample of value/duration pairs that may not capture abnormal events (i.e. outliers). If there is a desire to be made aware of long running updates, etc., I think we'd do well by having a trace event specifically for outliers and letting our more generic metric capture the typical rate.

Reply:
> You would just have a little uncertainty for one 5s window,

I thought Meng wanted to know the precise tLog ID.

> in particular is going to be better than a sample of value/duration pairs that may not capture abnormal events (i.e. outliers).

I thought the value/duration pairs would actually better capture the outliers (as you would get all the sample pints) and computing getRate() could lose that information (note that Counter::getRate() reports (interval_delta / (now() - interval_start)) and interval_data gets incremented when "+=" operator is called, so information about outliers could get averaged out).

Reply:
In both cases, this is only true if we log all of the data. If we log the most recent one every 5 seconds, then you are getting only a sample of the data and can't piece together the whole story.

Reply:
So, what is the current practice (or the recommended best practice) for monitoring an FDB cluster? Doesn't the app/user/tool pull status info from FoundationDB cluster at regular intervals, and stash and analyze that info for any potential issues?

Reply:
Some do use the data in status at regular interval to populate real-time dashboards for their clusters. You can also keep and query the source trace logs, which is common for doing more in-depth analysis.

Reply:
Ok, the counters are actually 5-second counters - I see that Counter::resetInterval() is getting called in traceCounters() and CounterCollection::logToTraceEvent(). I was under the impression that they were cumulative counters that would never get reset.

I also agree that either approach, counter based or <value, duration> pair, is not going to provide all the information that Meng wanted (like, is an SS not fetching versions because of tLog or network starvation).

> My gut says that having a rate is going to be more useful overall,

Shall we ask one of the SREs (who are the real consumers of these metrics) for an opinion as to which would be more useful? (I am fine either way now that I know that the counter is actually a 5-second counter and so would capture any latest changes in fetch behavior.) 

I do like to wrap up this issue soon so I can move on to my next task, so let's try to agree on the approach today.

Reply:
The questions we can answer with the two counter approach are:

1. How fast are we catching up in versions (i.e. are we going slower, equal to, or faster than the 1M/s rate that versions advance)
2. Is our fetching frequency lower than expected (a potential sign that something is wrong on the fetch path, such as log or network saturation)

Adding in the logging of outliers would give us some extra data about the magnitude of any weird occurrences. Assuming log fetches are sticky, we can also easily associate these pulls with a particular log, and I think we can then answer any of the questions that Meng had, at least to the degree that this particular metric is capable of answering them. If there are planned architecture changes such that log fetches are not sticky, then we'll need some additional metrics either way to track the behavior of SS/TL pairwise behavior.

> Shall we ask one of the SREs (who are the real consumers of these metrics) for an opinion as to which would be more useful?

For what it's worth, I am also a consumer of the metrics that FoundationDB produces, but I think getting buy in from your SRE team is a worthy endeavor.

Reply:
> For what it's worth, I am also a consumer of the metrics that FoundationDB produces, 

Sorry about my comment. I will take your word then. I am fine with the two counter approach. But I will wait for Meng to comment before making any coding changes.

Reply:
I'm fine with @sfc-gh-ajbeamon 's comment. We can add two counters to catch the two questions.